contracts/swap/swap.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package swap package wraps the 'swap' Ethereum smart contract.
+// It is an abstraction layer to hide implementation details about the different
+// Swap contract iterations (Swap, Swap, etc.)
+package swap
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethersphere/swarm/contracts/swap/contract"
+)
+
+var ErrTransactionReverted = errors.New("Transaction reverted")
+
+// Validator struct -> put validator in implementation of Swap. Make the validator a package level function and implement this in Swap
+
+// Backend wraps all methods required for contract deployment.
+type Backend interface {
+	bind.ContractBackend
+	TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error)
+	//TODO: needed? BalanceAt(ctx context.Context, address common.Address, blockNum *big.Int) (*big.Int, error)
+}
+
+// SimpleSwap interface defines the simple swap's exposed methods
+type SimpleSwap interface {
+	Deploy(auth *bind.TransactOpts, backend bind.ContractBackend, owner common.Address, harddepositTimeout *big.Int) (common.Address, *types.Transaction, error)
```

Comment:
Either we rename the smart contracts so that we don't need an abstraction in the code, or we do a proper proxying of all the contract methods, so that we don't need to have `Instance` public

Reply:
PR https://github.com/ethersphere/swarm/pull/1626 introduces a refactoring of all this.